src/newtondynamics_physics/SysNewton.cpp

```diff
@@ -76,75 +70,114 @@ using osp::Matrix3;
-    uintptr_t const userData = reinterpret_cast<uintptr_t>(
+    auto const userData = reinterpret_cast<uintptr_t>(
             NewtonBodyGetUserData(pBody));
-    ActiveEnt const ent = static_cast<ActiveEnt>(userData);
-
-    // Apply force
-
-    Vector3 force;
-    bool hasForce = false;
+    auto const bodyId = static_cast<BodyId>(userData);
 
-    Vector3 torque;
-    bool hasTorque = false;
+    Vector3 force{0.0f};
+    Vector3 torque{0.0f};
 
-    for (ACtxNwtWorld::ForceTorqueIn const& forceTorque
-         : pWorldCtx->m_forceTorqueIn)
+    auto factorBits = lgrn::bit_view(pWorldCtx->m_bodyFactors[bodyId]);
+    for (std::size_t const factorIdx : factorBits.ones())
     {
-        if (forceTorque.m_force.contains(ent))
-        {
-            force += forceTorque.m_force.get(ent);
-            hasForce = true;
-        }
+        ACtxNwtWorld::ForceFactorFunc const& factor = pWorldCtx->m_factors[factorIdx];
 
-        if (forceTorque.m_torque.contains(ent))
-        {
-            torque += forceTorque.m_torque.get(ent);
-            hasTorque = true;
-        }
+        factor.m_func(pBody, bodyId, *pWorldCtx, factor.m_userData, force, torque);
     }
 
-    if (hasForce)
-    {
-        NewtonBodySetForce(pBody, force.data());
-    }
-
-    if (hasTorque)
-    {
-        NewtonBodySetTorque(pBody, torque.data());
-    }
+    NewtonBodySetForce(pBody, force.data());
+    NewtonBodySetTorque(pBody, torque.data());
 } // cb_force_torque()
 
-static void cb_set_transform(NewtonBody const* const pBody,
+void SysNewton::cb_set_transform(NewtonBody const* const pBody,
                             const dFloat* const pMatrix, int threadIndex)
 {
     // Get context from Newton World
     ACtxNwtWorld* pWorldCtx = static_cast<ACtxNwtWorld*>(
             NewtonWorldGetUserData(NewtonBodyGetWorld(pBody)));
 
     // Get Associated entity
-    uintptr_t const userData = reinterpret_cast<uintptr_t>(NewtonBodyGetUserData(pBody));
-    ActiveEnt const ent = static_cast<ActiveEnt>(userData);
+    auto const userData = reinterpret_cast<uintptr_t>(NewtonBodyGetUserData(pBody));
+    auto const bodyId = static_cast<BodyId>(userData);
+    ActiveEnt const ent = pWorldCtx->m_bodyToEnt[bodyId];
 
-    pWorldCtx->m_perThread[threadIndex].m_setTf.emplace_back(ent, pBody);
+    NewtonBodyGetMatrix(pBody, pWorldCtx->m_pTransform->get(ent).m_transform.data());
 } // cb_set_transform()
 
-} // namespace ()
+
+void SysNewton::resize_body_data(ACtxNwtWorld& rCtxWorld)
+{
+    std::size_t const capacity = rCtxWorld.m_bodyIds.capacity();
+    rCtxWorld.m_bodyPtrs    .resize(capacity);
+    rCtxWorld.m_bodyToEnt   .resize(capacity);
+    rCtxWorld.m_bodyFactors .resize(capacity);
+}
+
+NewtonCollision* SysNewton::create_primative(
+        ACtxNwtWorld&   rCtxWorld,
+        EShape          shape)
+{
+    NewtonWorld* pNwtWorld = rCtxWorld.m_world.get();
+    NewtonCollision* pCollision;
+    switch (shape)
+    {
+    case EShape::Sphere:
```

Comment:
Minor: 

Consider the new (C++20) "using EShape;" syntax for switch-cases for readability.

This switch-case is small so the improvement is small.